### PR TITLE
new feature: Synchronous to asynchronous.when the namespace has too much items,loads lowly.Now it's 80 percent less time-consuming

### PR DESCRIPTION
**Pull request by renjiahua945**
1.Synchronous to asynchronous.when the namespace has too much items,loads lowly.Now it's 80 percent less time-consuming

2.Compact code.

## What's the purpose of this PR

In order to reduce the time consumption of NamespaceController#findNamespaces

## Which issue(s) this PR fixes:
Fixes #
NONE

## Brief changelog
1. Add a Spring task thread pool configuration
2. Modified NamespaceService#findNamespaceBOs，serial lines changed to parallel.
3. Add a  Class with name NamespaceBoAsyncService  to provider some future methon.

Follow this checklist to help us incorporate your contribution quickly and easily:

- [x] Read the [Contributing Guide](https://github.com/ctripcorp/apollo/blob/master/CONTRIBUTING.md) before making this pull request.
- [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [x] Write necessary unit tests to verify the code.
- [x] Run `mvn clean test` to make sure this pull request doesn't break anything.

### CODE REVIEW

**Comment by Anilople** [comment]
There are files conflict.

> This branch is 1 commit ahead, 114 commits behind ctripcorp:master.

Maybe you can [Syncing a fork](https://docs.github.com/en/free-pro-team@latest/github/collaborating-with-issues-and-pull-requests/syncing-a-fork) to make your branch to newest code of the project?

**Comment by renjiahua945** [comment]
> There are files conflict.
> 
> > This branch is 1 commit ahead, 114 commits behind ctripcorp:master.
> 
> Maybe you can [Syncing a fork](https://docs.github.com/en/free-pro-team@latest/github/collaborating-with-issues-and-pull-requests/syncing-a-fork) to make your branch to newest code of the project?

是最新的，可能触发了编辑器的代码格式化

**Comment by renjiahua945** [comment]
> > There are files conflict.
> > > This branch is 1 commit ahead, 114 commits behind ctripcorp:master.
> > 
> > 
> > Maybe you can [Syncing a fork](https://docs.github.com/en/free-pro-team@latest/github/collaborating-with-issues-and-pull-requests/syncing-a-fork) to make your branch to newest code of the project?
> 
> 今天fork的，不知道为啥不是最新的

重新提交了，本地可以编译过。

**Comment by Anilople** [comment]
根据CI的结果

apollo-portal模块存在test失败

```java
Tests in error: 
  NamespaceServiceTest.testFindNamespace:102 » NullPointer
```

在类`com.ctrip.framework.apollo.portal.service.NamespaceServiceTest`的第102行，出现了空指针

具体日志为
[6_JDK 8.txt](https://github.com/ctripcorp/apollo/files/5744144/6_JDK.8.txt)

**Comment by renjiahua945** [comment]
> 根据CI的结果
> 
> apollo-portal模块存在test失败
> 
> ```java
> Tests in error: 
>   NamespaceServiceTest.testFindNamespace:102 » NullPointer
> ```
> 
> 在类`com.ctrip.framework.apollo.portal.service.NamespaceServiceTest`的第102行，出现了空指针
> 
> 具体日志为
> [6_JDK 8.txt](https://github.com/ctripcorp/apollo/files/5744144/6_JDK.8.txt)
自动化测试脚本的问题，使用了线程池异步。未能得到。附上运行成功的截图，你能帮我看看自动化脚本怎么修改嘛，thanks
image

**Comment by renjiahua945** [comment]
> 根据CI的结果
> 
> apollo-portal模块存在test失败
> 
> ```java
> Tests in error: 
>   NamespaceServiceTest.testFindNamespace:102 » NullPointer
> ```
> 
> 在类`com.ctrip.framework.apollo.portal.service.NamespaceServiceTest`的第102行，出现了空指针
> 
> 具体日志为
> [6_JDK 8.txt](https://github.com/ctripcorp/apollo/files/5744144/6_JDK.8.txt)

添加了mock测试方法，已经解决

**Comment by renjiahua945** [comment]
> 根据CI的结果
> 
> apollo-portal模块存在test失败
> 
> ```java
> Tests in error: 
>   NamespaceServiceTest.testFindNamespace:102 » NullPointer
> ```
> 
> 在类`com.ctrip.framework.apollo.portal.service.NamespaceServiceTest`的第102行，出现了空指针
> 
> 具体日志为
> [6_JDK 8.txt](https://github.com/ctripcorp/apollo/files/5744144/6_JDK.8.txt)

谁review下代码

**Comment by nobodyiam** [comment]
I think the main change is the following 2 parts?

1. parallel executing `transformNamespace2BO`
2. parallel executing `loadLatestRelease`, `findItems` and `findDeletedItems`

I think the first part could speed up the loading times when there are many namespaces, but I'm not sure whether the second part could improve the performace a lot, as `loadLatestRelease` and `findDeletedItems` should be very fast.

With that said, I'm thinking whether we could simplify the changes, e.g. only keep the parallel executing `transformNamespace2BO` logic?

**Comment by renjiahua945** [comment]
> I think the main change is the following 2 parts?
> 
> 1. parallel executing `transformNamespace2BO`
> 2. parallel executing `loadLatestRelease`, `findItems` and `findDeletedItems`
> 
> I think the first part could speed up the loading times when there are many namespaces, but I'm not sure whether the second part could improve the performace a lot, as `loadLatestRelease` and `findDeletedItems` should be very fast.
> 
> With that said, I'm thinking whether we could simplify the changes, e.g. only keep the parallel executing `transformNamespace2BO` logic?

我在公司项目中实际使用，当一个namespace中有上千个个item，commit有2000多个item，这个findDeletedItems  100多ms以上，有时会超过400ms，通过异步能使耗时最多达到findDeletedItems 这么长时间，效果还是不错的。

**Comment by renjiahua945** [comment]
> @renjiahua945 #3258 修复了 findDeletedItems 接口慢的问题，可以一直到你本地环境看下效果。另外看下是否可以简化下 PR 的逻辑，比如是否不需要增加一个单独的异步 service，而是直接通过 Callable 搭配 ExecutorService 就能实现 loadLatestRelease 和 findItems 的并发调用。

使用Async更加方便，可以指定自己自定义的线程池。只需要在方法上加上@Async("executorName")即可。不添加异步service的话，Async这个会不生效。当然了，异步的service统一放该功能的异步方法也未尝不可，更方便阅读管理代码。simple不只是代码simple就简洁了对吧。当然我如果有时间我会继续贡献代码，将我使用中遇到的问题解决。thx

**Review by nobodyiam — COMMENTED** [review]
Basically it looks okay to me, but it would be better to add some unit test for NamespaceBoAsyncService even though its logic is simple

**Comment by renjiahua945** [comment]
> Basically it looks okay to me, but it would be better to add some unit test for NamespaceBoAsyncService even though its logic is simple

Ok，周末我加上。

**Comment by renjiahua945** [comment]
> Basically it looks okay to me, but it would be better to add some unit test for NamespaceBoAsyncService even though its logic is simple

添加了单元测试 thx

File: apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/service/NamespaceService.java [hunk 1]
```diff
@@ -25,51 +23,58 @@
...
+    private final UserInfoHolder userInfoHolder;
+    private final AdminServiceAPI.NamespaceAPI namespaceAPI;
+    private final ItemService itemService;
+    private final ReleaseService releaseService;
+    private final AppNamespaceService appNamespaceService;
+    private final InstanceService instanceService;
+    private final NamespaceBranchService branchService;
+    private final RolePermissionService rolePermissionService;
+    @Autowired
+    private NamespaceBoAsyncService namespaceBoAsyncService;
```
Comment (nobodyiam) [inline apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/service/NamespaceService.java:54]: Please inject this dependency as a constructor injection, as this is more friendly to unit test and is consistent with other dependencies.

File: apollo-portal/src/test/java/com/ctrip/framework/apollo/portal/service/NamespaceBoAsyncServiceTest.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    private String testClusterName = "default";
+    private String testNamespaceName = "application";
+    private Env testEnv = Env.DEV;
+
+    @Before
+    public void setup() {
+    }
+
+    @Test
+    public void testGetLatestReleaseAsync() throws ExecutionException, InterruptedException {
```
Comment (nobodyiam) [inline apollo-portal/src/test/java/com/ctrip/framework/apollo/portal/service/NamespaceBoAsyncServiceTest.java:44]: please add unit test to cover abnormal cases, e.g. when releaseService.loadLatestRelease throws exception, etc.

File: apollo-portal/src/test/java/com/ctrip/framework/apollo/portal/service/NamespaceBoAsyncServiceTest.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+                .thenReturn(someRelease);
+        Future<Map<String, String>> latestReleaseAsync = namespaceBoAsyncService.getLatestReleaseAsync(testAppId, testEnv, testClusterName, testNamespaceName);
+        Map<String, String> asyncRelease = latestReleaseAsync.get();
+        assertEquals(2, asyncRelease.size());
+        assertEquals("123", asyncRelease.get("a"));
+        assertEquals("123", asyncRelease.get("b"));
+    }
+
+    @Test
+    public void testGetItemsAsynce() throws ExecutionException, InterruptedException {
```
Comment (nobodyiam) [inline apollo-portal/src/test/java/com/ctrip/framework/apollo/portal/service/NamespaceBoAsyncServiceTest.java:57]: please add unit test to cover abnormal cases, e.g. when itemService.findItem throws exception, etc.

File: apollo-portal/src/test/java/com/ctrip/framework/apollo/portal/service/NamespaceServiceTest.java [hunk 4]
```diff
@@ -98,6 +108,10 @@ public void testFindNamespace() {
                                           testNamespaceName)).thenReturn(someRelease);
```
Comment (nobodyiam) [inline apollo-portal/src/test/java/com/ctrip/framework/apollo/portal/service/NamespaceServiceTest.java:108]: with the new logic, I think the following mock setup could be removed?

```java
    when(releaseService.loadLatestRelease(testAppId, Env.DEV, testClusterName,
                                          testNamespaceName)).thenReturn(someRelease);
    when(releaseService.loadLatestRelease(testAppId, Env.DEV, testClusterName, "hermes")).thenReturn(someRelease);
    when(itemService.findItems(testAppId, Env.DEV, testClusterName, testNamespaceName)).thenReturn(someItems);
```

**Review by nobodyiam — CHANGES_REQUESTED** [review]

File: apollo-portal/src/test/java/com/ctrip/framework/apollo/portal/service/NamespaceBoAsyncServiceTest.java (hunk 3 above)
Comment (nobodyiam) [inline apollo-portal/src/test/java/com/ctrip/framework/apollo/portal/service/NamespaceBoAsyncServiceTest.java:57]:
```suggestion
    public void testGetItemsAsync() throws ExecutionException, InterruptedException {
```

Comment (renjiahua945) [reply apollo-portal/src/test/java/com/ctrip/framework/apollo/portal/service/NamespaceBoAsyncServiceTest.java]: Now its ok

File: apollo-portal/src/test/java/com/ctrip/framework/apollo/portal/service/NamespaceServiceTest.java (hunk 4 above)
Comment (renjiahua945) [reply apollo-portal/src/test/java/com/ctrip/framework/apollo/portal/service/NamespaceServiceTest.java]: Removed

File: apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/service/NamespaceService.java (hunk 1 above)
Comment (renjiahua945) [reply apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/service/NamespaceService.java]: ok

File: apollo-portal/src/test/java/com/ctrip/framework/apollo/portal/service/NamespaceBoAsyncServiceTest.java (hunk 2 above)
Comment (renjiahua945) [reply apollo-portal/src/test/java/com/ctrip/framework/apollo/portal/service/NamespaceBoAsyncServiceTest.java]: Resolve

File: apollo-portal/src/test/java/com/ctrip/framework/apollo/portal/service/NamespaceBoAsyncServiceTest.java (hunk 3 above)
Comment (renjiahua945) [reply apollo-portal/src/test/java/com/ctrip/framework/apollo/portal/service/NamespaceBoAsyncServiceTest.java]: Resolve

**Comment by renjiahua945** [comment]
-----------------------------------------------------------------------------
Tests run: 8, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 5.135 sec <<< FAILURE! - in com.ctrip.framework.apollo.internals.RemoteConfigRepositoryTest
testLongPollingRefresh(com.ctrip.framework.apollo.internals.RemoteConfigRepositoryTest)  Time elapsed: 5.025 sec  <<< ERROR!
java.util.concurrent.TimeoutException: Waited 5000 milliseconds (plus 15 milliseconds, 85100 nanoseconds delay) for com.google.common.util.concurrent.SettableFuture@5e8f9e2d[status=PENDING]
	at com.google.common.util.concurrent.AbstractFuture.get(AbstractFuture.java:506)

File: apollo-portal/src/test/java/com/ctrip/framework/apollo/portal/service/NamespaceBoAsyncServiceTest.java [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+    private String testClusterName = "default";
+    private String testNamespaceName = "application";
+    private Env testEnv = Env.DEV;
+
+    @Before
+    public void setup() {
+    }
+
+    @Test
+    public void testGetLatestReleaseAsync() {
```
Comment (nobodyiam) [inline apollo-portal/src/test/java/com/ctrip/framework/apollo/portal/service/NamespaceBoAsyncServiceTest.java:44]: I don't think this unit test actually tests the exception cases, please take a look at the following code as a sample.

https://github.com/ctripcorp/apollo/blob/8613ee5c997a5f43761414ef8a4c8dd7ac6df9d9/apollo-portal/src/test/java/com/ctrip/framework/apollo/portal/spi/ctrip/CtripUserServiceTest.java#L153-L164

File: apollo-portal/src/test/java/com/ctrip/framework/apollo/portal/service/NamespaceBoAsyncServiceTest.java [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+            assertEquals(2, asyncRelease.size());
+            assertEquals("123", asyncRelease.get("a"));
+            assertEquals("123", asyncRelease.get("b"));
+        } catch (Exception e) {
+            assertTrue((e instanceof ExecutionException || e instanceof InterruptedException));
+        }
+    }
+
+    @Test
+    public void testGetItemsAsynce() {
```
Comment (nobodyiam) [inline apollo-portal/src/test/java/com/ctrip/framework/apollo/portal/service/NamespaceBoAsyncServiceTest.java:62]:
```suggestion
    public void testGetItemsAsync() {
```

File: apollo-portal/src/test/java/com/ctrip/framework/apollo/portal/service/NamespaceBoAsyncServiceTest.java (hunk 5 above)
Comment (nobodyiam) [reply apollo-portal/src/test/java/com/ctrip/framework/apollo/portal/service/NamespaceBoAsyncServiceTest.java:44]: Also we need to add the tests for normal case and abnormal cases, respectively.

Comment (renjiahua945) [reply apollo-portal/src/test/java/com/ctrip/framework/apollo/portal/service/NamespaceBoAsyncServiceTest.java:44]: 上层接口显式抛出的异常，如SearchUsersWithError() throws Exception，我需要您这个例子这样处理吗？好像不是一个情景。
显式抛出的异常这个方法不能用。只能catch或者 继续向上抛出。我翻遍了咱们所有的test，都没有

Comment (nobodyiam) [inline apollo-portal/src/test/java/com/ctrip/framework/apollo/portal/service/NamespaceBoAsyncServiceTest.java:44]: It doesn't smell right to catch the exception, in unit test we need to control the exceptions.

```suggestion
    public void testGetLatestReleaseAsync() throws ExecutionException, InterruptedException {
```

File: apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/service/NamespaceBoAsyncService.java [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+ * @date 2020/12/26
+ */
+
+@Service
+public class NamespaceBoAsyncService {
+    /**
+     * releaseService
+     */
+    @Autowired
+    private ReleaseService releaseService;
```
Comment (nobodyiam) [inline apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/service/NamespaceBoAsyncService.java:30]: other places has changed to constructor injection, I think we need th follow the same style here

File: apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/service/NamespaceService.java [hunk 8]
```diff
@@ -25,51 +23,59 @@
...
-    this.userInfoHolder = userInfoHolder;
-    this.namespaceAPI = namespaceAPI;
-    this.itemService = itemService;
-    this.releaseService = releaseService;
-    this.appNamespaceService = appNamespaceService;
-    this.instanceService = instanceService;
-    this.branchService = branchService;
-    this.rolePermissionService = rolePermissionService;
-  }
+    private final PortalConfig portalConfig;
```
Comment (nobodyiam) [inline apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/service/NamespaceService.java:43]: the code style looks different, please apply the apollo code style and reformat

File: apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/service/NamespaceService.java [hunk 9]
```diff
@@ -204,56 +208,59 @@ public NamespaceBO findPublicNamespaceForAssociatedNamespace(Env env, String app
...
-      ItemBO itemBO = transformItem2BO(itemDTO, releaseItems);
-
-      if (itemBO.isModified()) {
-        modifiedItemCnt++;
-      }
-
-      itemBOs.add(itemBO);
+    private NamespaceBO transformNamespace2BO(Env env, NamespaceDTO namespace) {
+        NamespaceBO namespaceBO = new NamespaceBO();
+        try {
```
Comment (nobodyiam) [inline apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/service/NamespaceService.java:213]: this try block is too big

File: apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/service/NamespaceService.java [hunk 10]
```diff
@@ -204,56 +208,59 @@ public NamespaceBO findPublicNamespaceForAssociatedNamespace(Env env, String app
...
+                    .filter(ItemBO::isModified)
+                    .count();
+            // Parse deleted items
+            List<ItemBO> deletedItemBos = parseDeletedItems(items, releaseItems, deletedItemsMap);
+            itemBOs.addAll(deletedItemBos);
+            namespaceBO.setItems(itemBOs);
+            // The number of items to be modified
+            namespaceBO.setItemModifiedCnt((int) modifiedItemCnt + deletedItemBos.size());
+        } catch (InterruptedException | ExecutionException e) {
+            logger.error("transformNamespace2BO error namespace:{}, env:{},err:{}",
```
Comment (nobodyiam) [inline apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/service/NamespaceService.java:258]: logger.error doesn't need to include the err in the message.

Comment (nobodyiam) [reply apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/service/NamespaceService.java:258]: Why do we swallow the exception here?

File: apollo-portal/src/test/java/com/ctrip/framework/apollo/portal/service/NamespaceBoAsyncServiceTest.java [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+                .thenReturn(someRelease);
+        Future<Map<String, String>> latestReleaseAsync = namespaceBoAsyncService.getLatestReleaseAsync(testAppId, testEnv, testClusterName, testNamespaceName);
+        try {
+            Map<String, String> asyncRelease = latestReleaseAsync.get();
+            assertNotNull(asyncRelease);
+            assertEquals(2, asyncRelease.size());
+            assertEquals("123", asyncRelease.get("a"));
+            assertEquals("123", asyncRelease.get("b"));
+        } catch (Exception e) {
+            assertTrue((e instanceof ExecutionException || e instanceof InterruptedException));
```
Comment (nobodyiam) [inline apollo-portal/src/test/java/com/ctrip/framework/apollo/portal/service/NamespaceBoAsyncServiceTest.java:57]: It looks a little weird to catch the exception here

File: apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/service/NamespaceBoAsyncService.java [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+    /**
+     * Get the latest release
+     *
+     * @param appId         appId
+     * @param env           env
+     * @param clusterName   cl
+     * @param namespaceName namespaceName
+     * @return Future<Map < String, String>>
+     */
+    @Async("queryRemoteExecutor")
```
Comment (nobodyiam) [inline apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/service/NamespaceBoAsyncService.java:46]: How can we test the async service? I think the ut only tests the services itself, not the async one

Comment (renjiahua945) [reply apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/service/NamespaceBoAsyncService.java:46]: 这个直接调用就是异步，可以获取线程名称，是不是queryRemoteExecutor-开头，

File: apollo-portal/src/test/java/com/ctrip/framework/apollo/portal/service/NamespaceBoAsyncServiceTest.java (hunk 11 above)
Comment (renjiahua945) [reply apollo-portal/src/test/java/com/ctrip/framework/apollo/portal/service/NamespaceBoAsyncServiceTest.java:57]: 这个就该直接抛出的，不该在这里测试的。

File: apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/service/NamespaceService.java (hunk 10 above)
Comment (renjiahua945) [reply apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/service/NamespaceService.java:258]: 抛出去上层处理吗

File: apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/service/NamespaceService.java (hunk 8 above)
Comment (renjiahua945) [reply apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/service/NamespaceService.java:43]: yes,但是我发现项目很多地方就是不统一，而且很乱。

File: apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/service/NamespaceBoAsyncService.java (hunk 12 above)
Comment (nobodyiam) [reply apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/service/NamespaceBoAsyncService.java]: I mean how can we test it in the unit test? It's now running in `main` thread.

Comment (renjiahua945) [reply apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/service/NamespaceBoAsyncService.java]: CLA 如何签署呢？ commit id和id指的是什么呢

Comment (nobodyiam) [reply apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/service/NamespaceBoAsyncService.java]: @renjiahua945 给 PR 留言下面的内容即可

I have read the CLA Document and I hereby sign the CLA

**Comment by renjiahua945** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by kezhenxu94** [comment]
recheck

File: apollo-portal/src/test/java/com/ctrip/framework/apollo/portal/service/NamespaceBoAsyncServiceTest.java (hunk 3 above)
Comment (nobodyiam) [reply apollo-portal/src/test/java/com/ctrip/framework/apollo/portal/service/NamespaceBoAsyncServiceTest.java]: Sorry but I don't see the unit test to cover abnormal cases, would you please take a look at this?

File: apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/service/NamespaceService.java (hunk 8 above)
Comment (nobodyiam) [reply apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/service/NamespaceService.java:43]: please format this file with the [apollo code style](https://github.com/ctripcorp/apollo/tree/master/apollo-buildtools/style)

File: apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/service/NamespaceService.java [hunk 13]
```diff
@@ -204,53 +221,57 @@ public NamespaceBO findPublicNamespaceForAssociatedNamespace(Env env, String app
...
+        if (!CollectionUtils.isEmpty(list)) {
+            items.addAll(list);
+        }
+        Map<String, String> releaseAsync = latestReleaseAsync.get();
+        if (Objects.nonNull(releaseAsync)) {
+            releaseItems.putAll(releaseAsync);
+        }
+    } catch (InterruptedException | ExecutionException e) {
+        logger.error("transformNamespace2BO error namespace:{}, env:{}",
+                namespace, env);
```
Comment (nobodyiam) [inline apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/service/NamespaceService.java:256]: How about
```suggestion
        logger.error("transformNamespace2BO error namespace:{}, env:{}",
                namespace, env, e);
```

File: apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/service/NamespaceBoAsyncService.java (hunk 12 above)
Comment (nobodyiam) [reply apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/service/NamespaceBoAsyncService.java]: > How can we test the async service? I think the ut only tests the services itself, not the async one

This question still remains here.

**Pull request closed without merging**